lib/grandpa/message_handler.go

```diff
@@ -427,7 +427,13 @@ func (s *Service) VerifyBlockJustification(hash common.Hash, justification []byt
 	}
 
 	if has {
-		return nil, fmt.Errorf("already have finalised block with setID=%d and round=%d", setID, fj.Round)
+		finalisedHeader, err := s.blockState.GetFinalisedHeader(fj.Round, setID)
```

Comment:
The function `GetFinalisedHeader` performs more operations (retrieve the hash, decode it, build the header struct) than the method `GetFinalisedHash` which basically does a database call to get the stored hash.
```suggestion
		storedFinalisedHash, err := s.blockState.GetFinalisedHash(fj.Round, setID)
		if err != nil {...}
		if storedFinalisedHash != hash {...}
```

Reply:
there is not function `s.blockState.GetFinalisedHash`

Reply:
That is because `s.blockState` is an interface that does not contains the method, but the implementation is here -> https://github.com/ChainSafe/gossamer/blob/9f19d317258c5f7d5128b3819ff2534dd4dbfbb9/dot/state/block_finalisation.go#L56

So, basically what you should do is introduce the method in the interface:

```go
type BlockState interface {
    ...
    GetFinalisedHash(round, setID uint64) (common.Hash, error)
}
```

and now you are able to use the method `s.blockState.GetFinalisedHash`